### PR TITLE
EZP-29238: As an Administrator I want to resize original images

### PR DESCRIPTION
**Pull request by kmadejski**
| Question           | Answer
| ------------------ | ------------------
| **JIRA issue**     | [EZP-29238](https://jira.ez.no/browse/EZP-29238)
| **Improvement**| yes
| **New feature**    | yes
| **Target version** | `master` 
| **BC breaks**      | no
| **Doc needed**     | yes

This PR provides the new command for resizing original images. It was initially created by Support as an example for the customer, but after that, the customer requested to add it to the product. 

The command helps to optimize storage usage by changing original images size and quality according to the selected variation.

**TODO**:
- [x] Implement feature / fix a bug.
- [x] Fix new code according to Coding Standards (`$ composer fix-cs`).
- [x] Ask for Code Review.

### CODE REVIEW

**Comment by thiagocamposviana** [comment]
In the original pull request there was a call to clear the global cache every loop:

`            unset(
                $GLOBALS[ 'eZContentObjectContentObjectCache' ],
                $GLOBALS[ 'eZContentObjectDataMapCache' ],
                $GLOBALS[ 'eZContentObjectVersionCache' ]
            );`

This has been done to avoid memory leaks when trying to regenerate the image alias for content types with dozen of thousands content.

Reducing the fetch limit was not enough because eZ keeps the cache of the fetched objects in those global variables.

**Comment by andrerom** [comment]
@thiagocamposviana This is a pure eZ Platform command. It can't know details about legacy like that, that is more for any kind of legacy handler or legacy bridge or legacy itself to handle.

**Comment by thiagocamposviana** [comment]
@andrerom I am just wondering if it has been tested for cases where you need to update the images for, lets say, 20,000 objects, does the new stack does not try to cache the data for objects like legacy? I am sure for legacy reducing the fetch limit per loop was not enough to make it work. Also when I was testing the original script I needed to make those changes, maybe because it was for an early version of ez platform, but it was necessary.

**Comment by andrerom** [comment]
> new stack does not try to cache the data for objects like legacy?

In v2 there should not be any direct in-memory cache like that at the moment in v1 it depends on Stash settings and doc advice you to disable in-memory caching for any long running CLI processes.

**Comment by thiagocamposviana** [comment]
Right, my case was for v1, so this might be good as is for v2.

**Comment by lserwatka** [comment]
@thiagocamposviana I would like to clarify. When I wrote, we (eZ) will take over, we will provide support for new stack only.

File: eZ/Bundle/EzPublishCoreBundle/Command/ResizeOriginalImagesCommand.php [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Output\OutputInterface;
+
+class ResizeOriginalImagesCommand extends ContainerAwareCommand
+{
+    const DEFAULT_ITERATION_COUNT = 25;
+
+    protected function configure()
+    {
+        $this->setName('ezplatform:resize-original-images')->setDefinition(
```
Comment (alongosz) [inline eZ/Bundle/EzPublishCoreBundle/Command/ResizeOriginalImagesCommand.php]: Hmm, there is a chance we'll create more commands related to image handling (like clearing image variation cache), so I'd rather call it `ezplatform:images:resize-original`. We also thought about extracting Image-related stuff (including Placeholder generators) to a separate bundle, so separating naming like this makes even more sense for the future.

Of course maybe my choice of naming is not the best, so suggestions welcomed (ping @adamwojs)

File: eZ/Bundle/EzPublishCoreBundle/Command/ResizeOriginalImagesCommand.php [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+        $searchService = $repository->getSearchService();
+        $filterManager = $this->getContainer()->get('liip_imagine.filter.manager');
+        $ioService = $this->getContainer()->get('ezpublish.fieldType.ezimage.io_service.published');
+        $extensionGuesser = $this->getContainer()->get('liip_imagine.extension_guesser');
+
+        $contentType = $contentTypeService->loadContentTypeByIdentifier($contentTypeIdentifier);
+        $query = new Query();
+        $query->filter = new Query\Criterion\ContentTypeIdentifier($contentType->identifier);
+
+        $totalCount = $repository->sudo(function () use ($query, $searchService) {
```
Comment (alongosz) [inline eZ/Bundle/EzPublishCoreBundle/Command/ResizeOriginalImagesCommand.php]: Please add an extra `user` option and set proper User Reference instead in the overridden `initialize` method (example could be found in some other command... ping me if needed).

File: eZ/Bundle/EzPublishCoreBundle/Command/ResizeOriginalImagesCommand.php [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+                }
+
+                $progressBar->advance();
+            }
+        }
+
+        $progressBar->finish();
+        $output->writeln('');
+        $output->writeln(sprintf("<info>All images has been successfully resized using '%s' variation.</info>",
+            $variation));
```
Comment (alongosz) [inline eZ/Bundle/EzPublishCoreBundle/Command/ResizeOriginalImagesCommand.php]: CS: this stands out on a diff, so maybe should be splitted into 4 lines (or more, I'd rely here on what PhpStorms would choose ;) ).

File: eZ/Bundle/EzPublishCoreBundle/Command/ResizeOriginalImagesCommand.php [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    }
+
+    /**
+     * Copy of eZ\Bundle\EzPublishCoreBundle\Imagine\IORepositoryResolver::store()
+     * We can't use original one because original method uses eZ\Bundle\EzPublishCoreBundle\Imagine\IORepositoryResolver::getFilePath()
+     * so we would end-up with image stored in _aliases instead of overwritten original image.
+     *
+     * @param IOServiceInterface $IOService
+     * @param BinaryInterface $binary
+     * @param Value $image
```
Comment (alongosz) [inline eZ/Bundle/EzPublishCoreBundle/Command/ResizeOriginalImagesCommand.php]: Nitpick: please use FQCN

File: eZ/Bundle/EzPublishCoreBundle/Command/ResizeOriginalImagesCommand.php (hunk 1 above)
Comment (adamwojs) [reply eZ/Bundle/EzPublishCoreBundle/Command/ResizeOriginalImagesCommand.php]: `ezplatform:images:resize-original` is fine for me

**Comment by kmadejski** [comment]
@alongosz all requested changes applied in https://github.com/ezsystems/ezpublish-kernel/pull/2357/commits/e1c68f64064bbd6963b190e513192ed984924b15.

**Comment by andrerom** [comment]
Seems the AppVeyor failure is unrelated _(php not installing, might be temprary)_, so @kmadejski I think you can send this to QA.

**Comment by kmadejski** [comment]
@mnocon updated in https://github.com/ezsystems/ezpublish-kernel/pull/2357/commits/c3124e0aa94b2c8d80f487267bbba96c51432aae with changes which you've requested.

File: eZ/Bundle/EzPublishCoreBundle/Command/ResizeOriginalImagesCommand.php [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+
+    protected function configure()
+    {
+        $this->setName('ezplatform:images:resize-original')->setDefinition(
+            [
+                new InputArgument('contentTypeIdentifier', InputArgument::REQUIRED,
+                    'Indentifier of ContentType which has ezimage FieldType.'),
+                new InputArgument('imageFieldIdentifier', InputArgument::REQUIRED,
+                    'Identifier of field of ezimage type.'),
+                new InputArgument('variation', InputArgument::OPTIONAL,
```
Comment (bdunogier) [inline eZ/Bundle/EzPublishCoreBundle/Command/ResizeOriginalImagesCommand.php]: I'd have expected those as options and not arguments.

Comment (kmadejski) [reply eZ/Bundle/EzPublishCoreBundle/Command/ResizeOriginalImagesCommand.php]: @bdunogier done in https://github.com/ezsystems/ezpublish-kernel/pull/2357/commits/aaf7b90621113cabc2624bc596e5acc3b1a2badd.

**Comment by kmadejski** [comment]
@mnocon I've added logic for re-publishing image content object to set proper width/height attributes in XML in https://github.com/ezsystems/ezpublish-kernel/pull/2357/commits/aaf7b90621113cabc2624bc596e5acc3b1a2badd. Test once again, please 🙂 

@alongosz / @adamwojs / @andrerom / @konradoboza since command has been significantly improved, please do the review again.

**Comment by lserwatka** [comment]
@mnocon is all good here?

**Comment by lserwatka** [comment]
@DominikaK we need a doc for this command.

**Pull request merged**